File: part_I/dijkstra_algorithm.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <limits>
#include <algorithm>
#include <unordered_map>
#include "grafo.hpp"

struct Node {
    int v;
    double dist;
    bool operator>(const Node& other) const {
        return dist > other.dist;
    }
};

std::unordered_map<int, double> dijkstra(const Grafo& grafo, int s, std::unordered_map<int, int>& antecessor) {
    int n = grafo.qtdVertices();
    std::unordered_map<int, double> dist;
    antecessor.clear();

    // Inicializa distâncias
    for (int v = 1; v <= n; v++) {
        dist[v] = std::numeric_limits<double>::infinity();
        antecessor[v] = -1;
    }
    dist[s] = 0;

    // Min-heap
    std::priority_queue<Node, std::vector<Node>, std::greater<Node>> pq;
    pq.push({s, 0});

    while (!pq.empty()) {
        Node atual = pq.top();
        pq.pop();
        int u = atual.v;

        // Relaxamento
        for (int viz : grafo.vizinhos(u)) {
            double peso = grafo.peso(u, viz);
            if (dist[u] + peso < dist[viz]) {
                dist[viz] = dist[u] + peso;
                antecessor[viz] = u;
                pq.push({viz, dist[viz]});
            }
        }
    }

    return dist;
}

// Reconstrói caminho de s até v usando antecessores
std::vector<int> reconstruirCaminho(int v, const std::unordered_map<int, int>& antecessor) {
    std::vector<int> caminho;
    for (int atual = v; atual != -1; atual = antecessor.at(atual)) {
        caminho.push_back(atual);
    }
    std::reverse(caminho.begin(), caminho.end());
    return caminho;
}

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Uso: " << argv[0] << " <arquivo_grafo> <vertice_inicial>" << std::endl;
        return 1;
    }

    std::string arquivo = argv[1];
    int s = std::stoi(argv[2]);

    Grafo grafo(arquivo);
    std::unordered_map<int, int> antecessor;
    auto dist = dijkstra(grafo, s, antecessor);

    // Imprime saída no formato pedido
    for (int v = 1; v <= grafo.qtdVertices(); v++) {
        std::cout << v << ": ";

        auto caminho = reconstruirCaminho(v, antecessor);
        for (size_t i = 0; i < caminho.size(); i++) {
            if (i > 0) std::cout << ",";
            std::cout << caminho[i];
        }

        std::cout << "; d=";
        if (dist[v] == std::numeric_limits<double>::infinity())
            std::cout << "inf";
        else
            std::cout << dist[v];
        std::cout << std::endl;
    }

    return 0;
}

File: part_I/python/dijkstra_algorithm.py
import heapq
import math
from typing import Dict, List, Tuple
import sys

# Assumindo que você tem uma classe Grafo equivalente
class Grafo:
    def __init__(self, arquivo: str):
        # Implementação da leitura do arquivo
        self.vertices = {}
        self._ler_arquivo(arquivo)
    
    def _ler_arquivo(self, arquivo: str):
        # Implemente a leitura do arquivo conforme sua necessidade
        with open(arquivo, 'r') as f:
            # Exemplo simples - adapte conforme seu formato
            lines = f.readlines()
            # Lógica para parse do arquivo
            pass
    
    def qtdVertices(self) -> int:
        return len(self.vertices)
    
    def vizinhos(self, u: int) -> List[int]:
        # Retorna lista de vértices vizinhos ao vértice u
        if u in self.vertices:
            return list(self.vertices[u].keys())
        return []
    
    def peso(self, u: int, v: int) -> float:
        # Retorna o peso da aresta entre u e v
        if u in self.vertices and v in self.vertices[u]:
            return self.vertices[u][v]
        return math.inf

class Node:
    def __init__(self, v: int, dist: float):
        self.v = v
        self.dist = dist
    
    def __lt__(self, other):
        return self.dist < other.dist
    
    def __gt__(self, other):
        return self.dist > other.dist

def dijkstra(grafo: Grafo, s: int) -> Tuple[Dict[int, float], Dict[int, int]]:
    n = grafo.qtdVertices()
    dist = {}
    antecessor = {}
    
    # Inicializa distâncias
    for v in range(1, n + 1):
        dist[v] = math.inf
        antecessor[v] = -1
    dist[s] = 0
    
    # Min-heap
    pq = []
    heapq.heappush(pq, Node(s, 0))
    
    while pq:
        atual = heapq.heappop(pq)
        u = atual.v
        
        # Relaxamento
        for viz in grafo.vizinhos(u):
            peso = grafo.peso(u, viz)
            if dist[u] + peso < dist[viz]:
                dist[viz] = dist[u] + peso
                antecessor[viz] = u
                heapq.heappush(pq, Node(viz, dist[viz]))
    
    return dist, antecessor

def reconstruir_caminho(v: int, antecessor: Dict[int, int]) -> List[int]:
    caminho = []
    atual = v
    while atual != -1:
        caminho.append(atual)
        atual = antecessor[atual]
    caminho.reverse()
    return caminho

def main():
    if len(sys.argv) != 3:
        print(f"Uso: {sys.argv[0]} <arquivo_grafo> <vertice_inicial>")
        sys.exit(1)
    
    arquivo = sys.argv[1]
    s = int(sys.argv[2])
    
    grafo = Grafo(arquivo)
    dist, antecessor = dijkstra(grafo, s)
    
    # Imprime saída no formato pedido
    for v in range(1, grafo.qtdVertices() + 1):
        print(f"{v}: ", end="")
        
        caminho = reconstruir_caminho(v, antecessor)
        print(",".join(map(str, caminho)), end="")
        
        print("; d=", end="")
        if dist[v] == math.inf:
            print("inf", end="")
        else:
            print(f"{dist[v]}", end="")
        print()

if __name__ == "__main__":
    main()
